### ISSUE TITLE
Ej 17 Guia 9

### ISSUE THREAD

**Issue by HoltzTomas** [PI-ITBA/2024_02#113]
Consulto si esta bien simplemente porque tuve que usar un ciclo dentro de cada llamado y tal vez no era la forma adecuada.

Pensé en hacerlo sumando `str[0] - '0'` en caso de que sea dígito y restando `-1` en caso de que sea alpha pero en casos como `"aaa2bb3"` no da el output deseado. Tampoco encontre otra forma de ir llevando registro si no era de esta forma.

```c
int bienformado(char *str) {
    // Caso base: si llegamos al final de la cadena, está bien formado
    if (*str == '\0') {
        return 0;
    }

    int letras = 0;
    while (isalpha(*str)) {
        letras++;
        str++;
    }

    if (!isdigit(*str)) {
        return 1; // no hay dígito después de las letras
    }

    int longitud = *str - '0'; // Convierte el dígito en un número
    if (letras != longitud) {
        return 1; // el número de letras no coincide con el dígito
    }

    return bienformado(str + 1);
}
```

**Comment by tomaspietravallo** [comment]
Hola! Esta mal hacerlo así con el ciclo. Lo resolviste de manera imperativa pero camuflado, cuando en el ciclo haces `str++` te estas avanzando en el string sin hacer los llamados recursivos. La idea es resolver todo de forma recursiva

**Comment by tomaspietravallo** [comment]
Estas validando el string "de a tramos" y trabajas imperativamente dentro de cada tramo. A modo de ejemplo si tomas el string es `"abc3"` estarías invocando a `bienFormado` con el puntero a `"abc3"` trabajas imperativamente hasta el final, y el proximo llamado recursivo ya es con `\0` y corta

**Comment by HoltzTomas** [comment]
Claro. La otra forma que se me ocurrió fue esta. Ahí si estaría bien?

```c
int bienformado(char* s) {
    if ( *s == '\0' ) {
        return 0;
    }
    if ( isalpha(*s) ) {
        return 1 + bienformado(s + 1);
    }
    else {
        int res = bienformado(s + 1);
        if ( !res ) {
            return -(*s - '0');
        }
        else {
            return 1; // Cualquier numero > 0
        }
    }
}
```

**Comment by tomaspietravallo** [comment]
Si! Ahora esta bien

**Comment by tomaspietravallo** [comment]
En estilo tenes algunos if/else que podrían ser if + return simplemente (no hace falta poner el `else` si estas retornando dentro de la rama condicional anterior) pero funcionamiento/ uso de recursividad esta todo bien
